File: ddd/tests/fixtures/error-contract/workspace/billing-domain/src/invoice.rs
//! Supported reference variants written directly, fully qualified, and through a
//! renamed import. The remaining two variants live in the child module `line`.

pub mod line;

use crate::errors::IssueInvoiceError;
use core::result::Result as Res;

/// Shares its spelling with `billing_use_case::invoice::Invoice`; the two
/// declarations are different symbols in different packages.
pub struct Invoice;

impl Invoice {
    pub fn issue(&mut self) -> Result<(), IssueInvoiceError> {
        Ok(())
    }
}

pub struct Qualified;

impl Qualified {
    pub fn issue(&mut self) -> core::result::Result<(), crate::errors::IssueInvoiceError> {
        Ok(())
    }
}

pub struct Renamed;

impl Renamed {
    pub fn issue(&mut self) -> Res<(), IssueInvoiceError> {
        Ok(())
    }
}

/// Concrete `Self` in an inherent impl, reached through the impl's owner.
pub struct Created;

impl Created {
    pub fn create() -> Result<Self, IssueInvoiceError> {
        Ok(Self)
    }
}

/// The same spelling appears in a doc comment and in a string literal, and the
/// operation itself declares no return type.
///
/// Historical note: this operation used to return `Result<(), IssueInvoiceError>`.
pub struct Undeclared;

impl Undeclared {
    pub fn issue(&mut self) {
        let _note = "Result<(), IssueInvoiceError>";
    }
}

File: ddd/tests/fixtures/error-contract/typescript-workspace/billing-domain/src/invoice.ts
/**
 * Supported reference variants written under the declared name, under an import
 * alias, and through a type-only import. The remaining two variants live in the
 * child module `invoice/line.ts`; this module is the named-file parent that owns
 * the directory beside it.
 */

import { IssueInvoiceError } from "./errors.ts";
import { IssueInvoiceError as RenamedInvoiceError } from "./errors.ts";
import type { IssueInvoiceError as TypeOnlyInvoiceError } from "./errors.ts";
import type { Result } from "./result.ts";

/**
 * Shares its spelling with `billing-use-case`'s own declarations; the two are
 * different symbols in different packages.
 */
export class Invoice {
  issue(): Result<void, IssueInvoiceError> {
    return { ok: true, value: undefined };
  }

  renamed(): Result<void, RenamedInvoiceError> {
    return { ok: true, value: undefined };
  }

  typeOnly(): Result<void, TypeOnlyInvoiceError> {
    return { ok: true, value: undefined };
  }
}
